File: dp/subsequence_subarray/different_subsequence/different_subsequence.go
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-12-10 11:56:26
 * @LastEditTime: 2022-12-10 12:23:50
 */
package differentsubsequence

/**
题目描述
给定一个字符串 s 和一个字符串 t ，计算在 s 的子序列中 t 出现的个数。
字符串的一个 子序列 是指，通过删除一些（也可以不删除）字符且不干扰剩余字符相对位置所组成的新字符串
（例如，"ACE" 是 "ABCDE" 的一个子序列，而 "AEC" 不是）
题目数据保证答案符合 32 位带符号整数范围。
*/
// 如果是连续的话，或许可以使用滑动窗口
func numDistinct(s string, t string) int {
	m, n := len(s), len(t)
	// dp[i][j]表示的是字符串s以i结尾的子序列中 t以j结尾出现的次数
	dp := make([][]int, m+1)
	for i, _ := range dp {
		dp[i] = make([]int, n+1)
	}

	// 当t=""时，空字符串是每个字符串的子序列
	for i := 0; i <= m; i++ {
		dp[i][0] = 1
	}

	// // 当s=""时，s中不存在任何的子序列
	// for i := 0; i <= n; i++ {
	// 	dp[0][i] = 0
	// }
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if s[i-1] == t[j-1] {
				// 这里比较难想到的点就在后面的dp[i-1][j]，case：s="bagg", t="bag"
				dp[i][j] = dp[i-1][j-1] + dp[i-1][j]
			} else {
				// 只能是在s上动
				dp[i][j] = dp[i-1][j]
			}
		}
	}

	return dp[m][n]
}

// 这里我们dp[i][j]的定义是s[0..i]和t[0..j]，也可以定义为s[i..]和t[j..]

File: dp/subsequence_subarray/different_subsequence/DifferentSubsequence.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-12-10 11:56:38
 * @LastEditTime: 2023-02-22 19:53:32
 */
package subsequence_subarray.different_subsequence;

/**
 * leetcode 115/剑指offerII-097 hard 不同的子序列
 * 
 * 给定一个字符串 s 和一个字符串 t ，计算在 s 的子序列中 t 出现的个数。
 * 字符串的一个 子序列 是指，通过删除一些（也可以不删除）字符且不干扰剩余字符相对位置所组成的新字符串
 * （例如，"ACE" 是 "ABCDE" 的一个子序列，而 "AEC" 不是）
 * 题目数据保证答案符合 32 位带符号整数范围。
 */
public class DifferentSubsequence{
    public int numDistinct(String s, String t) {
        int m = s.length(); 
        int n = t.length();

        // dp[i][j]表示的是t[0..j]在s[0..i]中出现的次数
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 0; i <= m; i++) {
            dp[i][0] = 1; // 空字符串出现一次
        }

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s.charAt(i - 1) == t.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j]; // 即使相等，也可能删除
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
            }
        }
        return dp[m][n];
    }

    // dp(i, j)表示的是t[j..]在s[i..]中出现的次数
    public int dp(String s, String t, int i, int j) {
        if (j == t.length()) {
            return 1;
        }

        if (i == s.length()) {
            return 0;
        }

        if (s.charAt(i) == t.charAt(j)) {
            return dp(s, t, i + 1, j + 1) + dp(s, t, i + 1, j);
        } else {
            return dp(s, t, i + 1, j);
        }
    }
}